File: Lab4/GO/Task2.go
// producer_consumer.go
package main

import (
	"fmt"
	"math/rand"
	"sync"
	"time"
)

const bufferSize = 5

var (
	buffer  = make([]int, bufferSize)
	in, out int
	mutex   sync.Mutex
	empty   = make(chan struct{}, bufferSize)
	full    = make(chan struct{}, bufferSize)
)

func produceItem() int {
	return rand.Intn(100)
}

func consumeItem(item int) {
	fmt.Printf("Consumed: %d\n", item)
}

func producer() {
	for i := 0; i < 10; i++ {
		item := produceItem()
		<-empty // espera espacio libre
		mutex.Lock()
		buffer[in] = item
		fmt.Printf("Produced: %d\n", item)
		in = (in + 1) % bufferSize
		mutex.Unlock()
		full <- struct{}{}
		time.Sleep(100 * time.Millisecond)
	}
}

func consumer() {
	for i := 0; i < 5; i++ {
		<-full // espera ítem disponible
		mutex.Lock()
		item := buffer[out]
		out = (out + 1) % bufferSize
		mutex.Unlock()
		empty <- struct{}{} // libera espacio
		consumeItem(item)
		time.Sleep(150 * time.Millisecond)
	}
}

func main() {
	rand.Seed(time.Now().UnixNano())

	// Inicializamos los espacios vacíos del buffer
	for i := 0; i < bufferSize; i++ {
		empty <- struct{}{}
	}

	go producer()
	go consumer()
	go consumer()

	time.Sleep(5 * time.Second)
}

File: Lab4/C/Task2.c
#include <stdio.h>
#include <stdlib.h>
#include <pthread.h>
#include <semaphore.h>
#include <unistd.h>

#define BUFFER_SIZE 5
int buffer[BUFFER_SIZE];
int in = 0, out = 0;

sem_t empty, full;
pthread_mutex_t mutex;

int produce_item() {
    return rand() % 100;
}

void consume_item(int item) {
    printf("Consumed: %d\n", item);
}

void *producer(void *arg) {
    for (int i = 0; i < 10; i++) {
        int item = produce_item();
        sem_wait(&empty);
        pthread_mutex_lock(&mutex);
        buffer[in] = item;
        printf("Produced: %d\n", item);
        in = (in + 1) % BUFFER_SIZE;
        pthread_mutex_unlock(&mutex);
        sem_post(&full);
        usleep(100000);
    }
    return NULL;
}

void *consumer(void *arg) {
    for (int i = 0; i < 10; i++) {
        sem_wait(&full);
        pthread_mutex_lock(&mutex);
        int item = buffer[out];
        out = (out + 1) % BUFFER_SIZE;
        pthread_mutex_unlock(&mutex);
        sem_post(&empty);
        consume_item(item);
        usleep(150000);
    }
    return NULL;
}

int main() {
    pthread_t prod, cons;
    sem_init(&empty, 0, BUFFER_SIZE);
    sem_init(&full, 0, 0);
    pthread_mutex_init(&mutex, NULL);

    pthread_create(&prod, NULL, producer, NULL);
    pthread_create(&cons, NULL, consumer, NULL);

    pthread_join(prod, NULL);
    pthread_join(cons, NULL);

    sem_destroy(&empty);
    sem_destroy(&full);
    pthread_mutex_destroy(&mutex);

    return 0;
}
